File: app/src/main/java/com/aayush/filmipedia/viewmodel/MovieCrewViewModel.java
package com.aayush.filmipedia.viewmodel;

import com.aayush.filmipedia.FilmipediaApplication;
import com.aayush.filmipedia.model.MovieCrew;
import com.aayush.filmipedia.util.NetworkState;
import com.aayush.filmipedia.util.datasource.MovieCrewDataSource;
import com.aayush.filmipedia.util.datasource.factory.MovieCrewDataFactory;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;
import io.reactivex.disposables.CompositeDisposable;

public class MovieCrewViewModel extends ViewModel {
    private LiveData<NetworkState> networkState;
    private LiveData<PagedList<MovieCrew>> movieCrewLiveData;
    private FilmipediaApplication application;
    private CompositeDisposable compositeDisposable;

    private Long movieId;

    public MovieCrewViewModel(@NonNull FilmipediaApplication application, Long movieId) {
        this.application = application;
        this.movieId = movieId;
        init();
    }

    private void init() {
        compositeDisposable = new CompositeDisposable();

        MovieCrewDataFactory movieCastDataFactory = new MovieCrewDataFactory(application,
                compositeDisposable, movieId);
        networkState = Transformations.switchMap(movieCastDataFactory.getMutableLiveData(),
                MovieCrewDataSource::getNetworkState);

        PagedList.Config pagedListConfig = new PagedList.Config.Builder()
                .setPageSize(10)
                .setInitialLoadSizeHint(20)
                .setEnablePlaceholders(false)
                .build();

        movieCrewLiveData = new LivePagedListBuilder<>(movieCastDataFactory, pagedListConfig)
                .build();

    }

    public LiveData<NetworkState> getNetworkState() {
        return networkState;
    }

    public LiveData<PagedList<MovieCrew>> getMovieCrewLiveData() {
        return movieCrewLiveData;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.dispose();
    }
}

File: app/src/main/java/com/aayush/filmipedia/viewmodel/MovieCrewViewModel.kt
package com.aayush.filmipedia.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.aayush.filmipedia.FilmipediaApplication
import com.aayush.filmipedia.model.MovieCrew
import com.aayush.filmipedia.util.NetworkState
import com.aayush.filmipedia.util.datasource.factory.MovieCrewDataFactory
import io.reactivex.disposables.CompositeDisposable

class MovieCrewViewModel(application: FilmipediaApplication, movieId: Long?) : ViewModel() {
    var networkState: LiveData<NetworkState>
        private set
    var movieCrewLiveData: LiveData<PagedList<MovieCrew>>
        private set
    private var compositeDisposable = CompositeDisposable()

    init {
        val movieCastDataFactory = MovieCrewDataFactory(application, compositeDisposable, movieId)
        networkState = Transformations.switchMap(movieCastDataFactory.mutableLiveData) { it.networkState }

        val pagedListConfig = PagedList.Config.Builder()
            .setPageSize(10)
            .setInitialLoadSizeHint(20)
            .setEnablePlaceholders(false)
            .build()

        movieCrewLiveData = LivePagedListBuilder(movieCastDataFactory, pagedListConfig)
            .build()
    }

    override fun onCleared() {
        super.onCleared()
        compositeDisposable.dispose()
    }
}